api/core/features/assistant_fc_runner.py

```diff
@@ -0,0 +1,335 @@
+      SystemPromptMessage, AssistantPromptMessage, ToolPromptMessage, PromptMessageTool
+from core.model_runtime.entities.llm_entities import LLMResultChunk, LLMResult, LLMUsage
+from core.model_manager import ModelInstance
+from core.application_queue_manager import PublishFrom
+
+from core.tools.errors import ToolInvokeError, ToolNotFoundError, \
+    ToolNotSupportedError, ToolProviderNotFoundError, ToolParamterValidationError, \
+          ToolProviderCredentialValidationError
+
+from core.features.assistant_base_runner import BaseAssistantApplicationRunner
+
+from models.model import Conversation, Message, MessageAgentThought
+
+logger = logging.getLogger(__name__)
+
+class AssistantFunctionCallApplicationRunner(BaseAssistantApplicationRunner):
+    def run(self, model_instance: ModelInstance,
+                conversation: Conversation,
+                message: Message,
+                query: str,
+    ) -> Generator[LLMResultChunk, None, None]:
+        """
+        Run FunctionCall agent application
+        """
+        app_orchestration_config = self.app_orchestration_config
+
+        prompt_template = self.app_orchestration_config.prompt_template.simple_prompt_template or ''
+        prompt_messages = self.history_prompt_messages
+        prompt_messages = self.organize_prompt_messages(
+            prompt_template=prompt_template,
+            query=query,
+            prompt_messages=prompt_messages
+        )
+
+        # convert tools into ModelRuntime Tool format
+        prompt_messages_tools: List[PromptMessageTool] = []
+        tool_instances = {}
+        for tool in self.app_orchestration_config.agent.tools if self.app_orchestration_config.agent else []:
+            try:
+                prompt_tool, tool_entity = self._convert_tool_to_prompt_message_tool(tool)
+            except Exception:
+                # api tool may be deleted
+                continue
+            # save tool entity
+            tool_instances[tool.tool_name] = tool_entity
+            # save prompt tool
+            prompt_messages_tools.append(prompt_tool)
+
+        # convert dataset tools into ModelRuntime Tool format
+        for dataset_tool in self.dataset_tools:
+            prompt_tool = self._convert_dataset_retriever_tool_to_prompt_message_tool(dataset_tool)
+            # save prompt tool
+            prompt_messages_tools.append(prompt_tool)
+            # save tool entity
+            tool_instances[dataset_tool.identity.name] = dataset_tool
+
+        iteration_step = 1
+        max_iteration_steps = min(app_orchestration_config.agent.max_iteration, 5) + 1
+
+        # continue to run until there is not any tool call
+        function_call_state = True
+        agent_thoughts: List[MessageAgentThought] = []
+        llm_usage = {
+            'usage': None
+        }
+        final_answer = ''
+
+        def increse_usage(final_llm_usage_dict: Dict[str, LLMUsage], usage: LLMUsage):
```

Comment:
did you mean increase?
```suggestion
        def increase_usage(final_llm_usage_dict: Dict[str, LLMUsage], usage: LLMUsage):
```

Reply:
That's it! thanks for your suggestions! 😊